### PR TITLE
Fix `CairoPie` serialization.

### PR DESCRIPTION
**Pull request by azteca1998**
# Fix `CairoPie` serialization

## Description

See #1443 for details.

## Checklist
- [x] Linked to Github Issue
- [x] Unit tests added
- [ ] Integration tests added.
- [ ] This change requires new documentation.
  - [ ] Documentation has been added/updated.
  - [x] CHANGELOG has been updated.

### CODE REVIEW

**Review by Oppen — APPROVED** [review]

**Review by Oppen — APPROVED** [review]

File: CHANGELOG.md [hunk 1]
```diff
@@ -107,7 +109,7 @@
     `get_hint_data(self, &[HintReference], &mut dyn HintProcessor) -> Result<Vec<Box<dyn Any>, VirtualMachineError>`
   * Hook methods receive `&[Box<dyn Any>]` rather than `&HashMap<usize, Vec<Box<dyn Any>>>`
 
-#### [0.8.4]
```
Comment (pefontana) [inline CHANGELOG.md:110]: Can we restore these unnecessary changes?

File: vm/src/serde/deserialize_program.rs
Comment (pefontana) [inline vm/src/serde/deserialize_program.rs:1]: Same, restore the changes in this file

File: vm/src/tests/.gitignore
Comment (pefontana) [inline vm/src/tests/.gitignore:1]: Isn't this already handled in the `./.gitignore`?
In case is not, add this line to the  `./.gitignore` instead of creating a new file

File: vm/src/tests/cairo_pie_test.rs [hunk 2]
```diff
@@ -1,22 +1,9 @@
...
-
-#[cfg(target_arch = "wasm32")]
-use wasm_bindgen_test::*;
-
-#[cfg(all(not(feature = "std"), feature = "alloc"))]
-use alloc::{
-    string::{String, ToString},
-    vec::Vec,
-};
+#![deny(warnings)]
```
Comment (pefontana) [inline vm/src/tests/cairo_pie_test.rs:1]: Remove the #![deny(warnings)]

File: vm/src/tests/cairo_pie_test.rs [hunk 3]
```diff
@@ -269,7 +266,11 @@ fn serialize_and_deserialize_cairo_pie() {
     let result = runner.get_cairo_pie(&vm);
     assert!(result.is_ok());
     let cairo_pie = result.unwrap();
-    let cairo_pie_serialized = serde_json::to_string(&cairo_pie).unwrap();
-    let cairo_pie_deserialized: CairoPie = serde_json::from_str(&cairo_pie_serialized).unwrap();
-    assert_eq!(cairo_pie_deserialized, cairo_pie);
+
+    // TODO: Find a way to test the entire object, and not just `metadata`.
```
Comment (pefontana) [inline vm/src/tests/cairo_pie_test.rs:270]: Aren´t we checking the other fields, like `memory`, `additional_data`. etc

File: vm/src/vm/runners/cairo_pie.rs [hunk 4]
```diff
@@ -52,15 +54,17 @@ pub enum BuiltinAdditionalData {
...
 
-#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, Eq)]
+#[derive(Serialize, Clone, Debug, PartialEq, Eq)]
 pub struct CairoPie {
     pub metadata: CairoPieMetadata,
+    #[serde(serialize_with = "serde_impl::serialize_memory")]
     pub memory: CairoPieMemory,
     pub execution_resources: ExecutionResources,
     pub additional_data: HashMap<String, BuiltinAdditionalData>,
+    pub version: CairoPieVersion,
```
Comment (pefontana) [inline vm/src/vm/runners/cairo_pie.rs:64]: I don't know if it is possible, but can we use a String type for the version field.
Plus add a constant with the CairoPieVersion String, so it is easy to change

**Review by pefontana — CHANGES_REQUESTED** [review]
Nice @azteca1998 !
I think that with some little changes we will be good to go

File: CHANGELOG.md (hunk 1 above)
Comment (azteca1998) [reply CHANGELOG.md:110]: This is done automatically by the editor. It removes any trailing spaces when saving the file.

File: vm/src/vm/runners/cairo_pie.rs (hunk 4 above)
Comment (azteca1998) [reply vm/src/vm/runners/cairo_pie.rs:64]: While it is possible, it'd need to be initialized every time the CairoPie object is created. Having it as it is now this cannot happen because the version field is zero-sized.

The constant is easy. Fixed.

File: vm/src/tests/cairo_pie_test.rs (hunk 3 above)
Comment (azteca1998) [reply vm/src/tests/cairo_pie_test.rs:270]: Yes, I forgot to remove it. Fixed, thanks.

File: vm/src/tests/.gitignore
Comment (azteca1998) [reply vm/src/tests/.gitignore:1]: Fixed.

File: vm/src/tests/cairo_pie_test.rs (hunk 2 above)
Comment (azteca1998) [reply vm/src/tests/cairo_pie_test.rs]: Fixed.

File: vm/src/serde/deserialize_program.rs
Comment (azteca1998) [reply vm/src/serde/deserialize_program.rs:1]: Fixed.

File: vm/src/vm/runners/cairo_pie.rs (hunk 4 above)
Comment (pefontana) [reply vm/src/vm/runners/cairo_pie.rs:66]: nice

**Review by pefontana — APPROVED** [review]
